### ISSUE TITLE
HMAC_SHA256 conflict with built-in library

### ISSUE THREAD

**Issue by sqmsmu** [Ant2000/CustomJWT#6]
While trying to build for esp32-s3 in Platformio using arduino framework, I am getting the following error

Linking .pio/build/esp32-s3-devkitc-1/firmware.elf
/Users/muddassirsyed/.platformio/packages/toolchain-xtensa-esp32s3/bin/../lib/gcc/xtensa-esp32s3-elf/8.4.0/../../../../xtensa-esp32s3-elf/bin/ld: /Users/muddassirsyed/.platformio/packages/framework-arduinoespressif32/tools/sdk/esp32s3/lib/libwpa_supplicant.a(crypto_mbedtls.c.obj): in function `hmac_sha256':
/Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/wpa_supplicant/src/crypto/crypto_mbedtls.c:258: multiple definition of `hmac_sha256'; .pio/build/esp32-s3-devkitc-1/lib176/libCustomJWT.a(hmac_sha256.c.o):/Users/muddassirsyed/Neuska/git/NeuskaPortal/Heimdall/heimdall/.pio/libdeps/esp32-s3-devkitc-1/CustomJWT/src/hmac_sha256.c:36: first defined here

I think it is conflicting with this change "Implemented a standalone implementation of HMAC-SHA256", v2.1.0 works fine for the build.

**Comment by jo-russ** [comment]
> While trying to build for esp32-s3 in Platformio using arduino framework, I am getting the following error
> 
> Linking .pio/build/esp32-s3-devkitc-1/firmware.elf /Users/muddassirsyed/.platformio/packages/toolchain-xtensa-esp32s3/bin/../lib/gcc/xtensa-esp32s3-elf/8.4.0/../../../../xtensa-esp32s3-elf/bin/ld: /Users/muddassirsyed/.platformio/packages/framework-arduinoespressif32/tools/sdk/esp32s3/lib/libwpa_supplicant.a(crypto_mbedtls.c.obj): in function `hmac_sha256': /Users/ficeto/Desktop/ESP32/ESP32S2/esp-idf-public/components/wpa_supplicant/src/crypto/crypto_mbedtls.c:258: multiple definition of `hmac_sha256'; .pio/build/esp32-s3-devkitc-1/lib176/libCustomJWT.a(hmac_sha256.c.o):/Users/muddassirsyed/Neuska/git/NeuskaPortal/Heimdall/heimdall/.pio/libdeps/esp32-s3-devkitc-1/CustomJWT/src/hmac_sha256.c:36: first defined here
> 
> I think it is conflicting with this change "Implemented a standalone implementation of HMAC-SHA256", v2.1.0 works fine for the build.

Same boat with esp32-cam. Here's what I did and it compiled without problems.
Basically rename anything hmac_sha256 to [something-unique]_hmac_sha256 I have chosen library name as a safe bet.

```
-------------------8<----------------8<----------------8<----------------8<----------------8<------
 src/CustomJWT.h   | 2 +-
 src/hmac_sha256.c | 2 +-
 src/hmac_sha256.h | 8 ++++----
 3 files changed, 6 insertions(+), 6 deletions(-)

diff --git a/src/CustomJWT.h b/src/CustomJWT.h
index 2aac76a..b383cf0 100644
--- a/src/CustomJWT.h
+++ b/src/CustomJWT.h
@@ -209,7 +209,7 @@ public:
     {
         uint8_t hashed[SHA256_HASH];
         memset(hashed, 0, SHA256_HASH);
-        hmac_sha256(secret, secretLen, data, dataLen, hashed, SHA256_HASH);
+        customjwt_hmac_sha256(secret, secretLen, data, dataLen, hashed, SHA256_HASH);
         Base64URL::base64urlEncode(hashed, SHA256_HASH, output, outputLen);
     }
 
diff --git a/src/hmac_sha256.c b/src/hmac_sha256.c
index 5f7c0f8..67261ee 100644
--- a/src/hmac_sha256.c
+++ b/src/hmac_sha256.c
@@ -28,7 +28,7 @@ static void* sha256(const void* data,
                     const size_t outlen);
 
 // Declared in hmac_sha256.h
-size_t hmac_sha256(const void* key,
+size_t customjwt_hmac_sha256(const void* key,
                    const size_t keylen,
                    const void* data,
                    const size_t datalen,
diff --git a/src/hmac_sha256.h b/src/hmac_sha256.h
index 21f7086..719673c 100644
--- a/src/hmac_sha256.h
+++ b/src/hmac_sha256.h
@@ -3,8 +3,8 @@
     Originally written by https://github.com/h5p9sl
 */
 
-#ifndef _HMAC_SHA256_H_
-#define _HMAC_SHA256_H_
+#ifndef _CUSTOMJWT_HMAC_SHA256_H_
+#define _CUSTOMJWT_HMAC_SHA256_H_
 
 #ifdef __cplusplus
 extern "C" {
@@ -13,7 +13,7 @@ extern "C" {
 #include <stddef.h>
 
 size_t  // Returns the number of bytes written to `out`
-hmac_sha256(
+customjwt_hmac_sha256(
     // [in]: The key and its length.
     //      Should be at least 32 bytes long for optimal security.
     const void* key,
@@ -33,4 +33,4 @@ hmac_sha256(
 }
 #endif  // __cplusplus
 
-#endif  // _HMAC_SHA256_H_
+#endif  // _CUSTOMJWT_HMAC_SHA256_H_
```

**Comment by Ant2000** [comment]
Thanks a lot for raising this issue. I have made a fix for this in the release v2.1.2 as per what @jo-russ described.
